src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/IProjectTreeExtensions.cs

```diff
@@ -12,120 +10,13 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS
-            Func<string, bool> itemFilter,
-            Func<string, IProjectTree> nodeCreator)
-        {
-            foreach (string item in itemsToAdd)
-            {
-                if (itemFilter(item))
-                {
-                    // Double check we don't already have this one -  we shouldn't
-                    IProjectTree existingNode = tree.FindImmediateChildByPath(item);
-                    System.Diagnostics.Debug.Assert(existingNode == null);
-                    if (existingNode == null)
-                    {
-                        tree = tree.Add(nodeCreator(item)).Parent;
-                    }
-                }
-            }
-            return tree;
-        }
-
-        /// <summary>
-        /// Removes given list of elements form IProjectTree
-        /// </summary>
-        /// <param name="tree"></param>
-        /// <param name="itemsToRemove"></param>
-        /// <param name="itemFilter"></param>
-        /// <returns></returns>
-        public static IProjectTree RemoveElementsFromTree(this IProjectTree tree,
-            IEnumerable<string> itemsToRemove,
-            Func<string, bool> itemFilter)
-        {
-            foreach (string file in itemsToRemove)
-            {
-                if (itemFilter(file))
-                {
-                    // See if we already have this one -  we shouldn't
-                    IProjectTree existingNode = tree.FindImmediateChildByPath(file);
-                    System.Diagnostics.Debug.Assert(existingNode != null);
-                    if (existingNode != null)
-                    {
-                        tree = existingNode.Remove();
-                    }
-                }
-            }
-            return tree;
-        }
-
-        /// <summary>
-        ///     Finds the node with the specified path within the specified node's 
-        ///     immediate children.
-        /// </summary>
-        /// <param name="tree">
-        ///     The <see cref="IProjectTree"/> to search
-        /// </param>
-        /// <param name="path">
-        ///     The path of the child to find.
-        /// </param>
-        /// <returns>
-        ///     The <see cref="IProjectTree"/> with the specified <paramref name="path"/>, 
-        ///     otherwise, <see langword="null"/> if the node could not be found.
-        /// </returns>
-        /// <exception cref="ArgumentNullException">
-        ///     <paramref name="tree"/> is <see langword="null"/>.
-        ///     <para>
-        ///         -or-
-        ///     </para>
-        ///     <paramref name="path"/> is <see langword="null"/>.
-        /// </exception>
-        /// <exception cref="ArgumentException">
-        ///     <paramref name="path"/> is an empty string ("")
-        /// </exception>
-        public static IProjectTree FindImmediateChildByPath(this IProjectTree tree, string path)
+        public static IProjectTree GetChildWithCaption(this IProjectTree tree, string caption)
```

Comment:
Interesting that you changed this to "Get"; across the framework we tend to use "Find" to indicate an API that can return null, such as this one.

Reply:
And CPS uses FindXXX/TryFindXXX, I would favor it.